### PR TITLE
Defined built-in kernels (DBK) extension with infrastructure for more drivers

### PR DESCRIPTION
**Pull request by franz**
This fixes a bunch of problems in the original PR (#1503).

- adds proper CMake support for libxsmm
- builtin_kernels.cc: converted from C++ to C
- simplified the `cl_tensor_desc` struct
- removed `custom_kernel_runner`, use existing paths for execution
- implemented a more structured API closer (but not identical) to the draft document on the extension
- extended CPU support of GEMM DBK to types other than FP32

### CODE REVIEW

**Review by pjaaskel — CHANGES_REQUESTED** [review]
This is now looking quite good. Only small ones. @TopiLeppanen please take a look at this PR from the perspective of FPGA/AlmaIF implementation before we pull this in and test that it doesn't break your work? @linehill can you update the extension specs to match this prototype and point from the extension PR to this PR?

File: include/CL/cl_exp_tensor.h [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+{
+  cl_char c;
+  cl_short s;
+  cl_int i;
+  cl_long l;
+  cl_half h;
+  cl_float f;
+  cl_double d;
+  char raw[sizeof (cl_double)];
+} cl_tensor_datatype_union;
```
Comment (linehill) [inline include/CL/cl_exp_tensor.h:75]: What if we need to pass objects larger than `sizeof(cl_double)` in the future? We can’t update this union due to potential ABI breakage.

File: lib/CL/clCreateProgramWithDefinedBuiltInKernels.c [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+ */
+
+CL_API_ENTRY cl_program CL_API_CALL
+POname (clCreateProgramWithDefinedBuiltInKernels) (
+  cl_context context,
+  cl_uint num_devices,
+  const cl_device_id *device_list,
+  cl_uint num_kernels,
+  const BuiltinKernelId *kernel_ids,
+  const char **kernel_names,
```
[the remark this answers is not in the document]
Comment (franz) [reply lib/CL/clCreateProgramWithDefinedBuiltInKernels.c:47]: added a check for uniqueness

File: include/CL/cl_exp_defined_builtin_kernels.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  CL_DBK_PROPERTY_NON_DETERMINISTIC,
+
+  // Allow driver to trade off accuracy for speed by allowing it to flush
+  // denormals to zero.
+  //
+  // Drivers may ignore this property, meaning the behavior is not guaranteed.
+  CL_DBK_PROPERTY_ALLOW_FTZ
+};
+
+typedef cl_kernel (CL_API_CALL *clCreateBuiltinKernelWithAttributesEXP_fn) (
```
[the remark this answers is not in the document]
Comment (franz) [reply include/CL/cl_exp_defined_builtin_kernels.h]: "clCreateProgramWithDefinedBuiltInKernels" is the only new API function in this PR

**Comment by TopiLeppanen** [comment]
Hello, I fixed some breakage in the AlmaIF driver related to the builtin_kernels.hh being converted to C.
Hope you don't mind I pushed it already here, since that seemed most convenient.

There was a missing pocl_BIDescriptor for the first built-in kernel in the list: pocl.copy.i8 which I added.

**Pull request merged**